### PR TITLE
[`Llama2`] Add disabling TP behavior

### LINKED ISSUE #726: Support Tensor Parallelism, which is used in LLaMA-2
### Feature request

[`transformers`'s LLaMA-2 implementation](https://github.com/huggingface/transformers/pull/24891) can bypass adapters because it directly split linear weight as `self.q_proj.weight.split` when Tensor Parallelism is enabled. Ideally, adapters should be split accordingly

https://github.com/huggingface/transformers/blob/476be08c4aa96f8c1cae4200d2677bbe8f12cf80/src/transformers/models/llama/modeling_llama.py#L291

### Motivation

To support LLaMA-2

### Your contribution

Discussion, reviewing a PR or submitting a PR

### PR DESCRIPTION
**Pull request by younesbelkada**
Fixes #726 

This PR is on par with https://github.com/huggingface/transformers/pull/24906

Currently the TP paradigm that is supported by transformers, technically is not really a real Tensor Parallelism paradigm but rather a simulation of TP by manually splitting the layers into chunks and concatenating the results.

The motivation of that implementation is to mimic the TP paradigm that was used during pre-training of these models, as slicing weight tensors and input leads to slight numerical differences: https://github.com/pytorch/pytorch/issues/76232 

I would argue that this might be not that important for training as the model will be fine-tuned, thus the weights of the model will be adapted accordingly. 

cc @pacman100 @BenjaminBossan 

I propose to properly support TP in the future once this is properly implemented, currently the TP that is in place is more a patch to match the logits of the original implementation

### CODE REVIEW

**Review by pacman100 — APPROVED** [review]
Thank you @younesbelkada for the quick fix, LGTM!

**Review by BenjaminBossan — APPROVED** [review]
It sounds reasonable to me that for fine-tuning, TP is disabled, especially if it is just simulated. I wonder if this should be documented somewhere, since, as you mentioned, it can lead to small numerical differences. Perhaps a comment above these new lines of code?

**Comment by younesbelkada** [comment]
> Perhaps a comment above these new lines of code?

Sure yes, will add this !
